auth/auth/auth.py

```diff
@@ -381,27 +487,86 @@ async def post_create_user(request, userdata):  # pylint: disable=unused-argumen
 
-    if login_id == '':
-        if not is_service_account:
-            set_message(session, 'Login id is required for users that are not service accounts.', 'error')
-            return web.HTTPFound(deploy_config.external_url('auth', '/users'))
-        login_id = None
+    if existing_user is None:
+        if login_id == '':
+            login_id = None
 
-    user_id = await db.execute_insertone(
+        user_id = await db.execute_insertone(
+            '''
+    INSERT INTO users (state, username, login_id, is_developer, is_service_account)
+    VALUES (%s, %s, %s, %s, %s);
+    ''',
+            ('creating', username, login_id, is_developer, is_service_account),
+        )
+        set_message(session, f'Created user {user_id} {username} {login_id}.', 'info')
+    else:
+        user_id = existing_user['id']
+        set_message(session, f'User {user_id} {username} {login_id} already exists.', 'info')
+
+    return web.HTTPFound(deploy_config.external_url('auth', '/users'))
+
+
+@routes.get('/api/v1alpha/users')
+@rest_authenticated_developers_only
+async def rest_get_users(request, userdata):  # pylint: disable=unused-argument
+    db: Database = request.app['db']
+    users = await db.select_and_fetchall(
         '''
-INSERT INTO users (state, username, login_id, is_developer, is_service_account)
-VALUES (%s, %s, %s, %s, %s);
+SELECT id, username, login_id, state, is_developer, is_service_account FROM users;
+'''
+    )
+    return web.json_response([user async for user in users])
+
+
+@routes.get('/api/v1alpha/users/{user}')
+@rest_authenticated_developers_only
+async def rest_get_users(request, userdata):  # pylint: disable=unused-argument
+    db: Database = request.app['db']
+    username = request.match_info['user']
+
+    user = await db.select_and_fetchone(
+        '''
+SELECT id, username, login_id, state, is_developer, is_service_account FROM users
+WHERE username = %s;
 ''',
-        ('creating', username, login_id, is_developer, is_service_account),
+        (username,),
     )
+    return web.json_response(user)
 
-    set_message(session, f'Created user {user_id} {username} {login_id}.', 'info')
 
-    return web.HTTPFound(deploy_config.external_url('auth', '/users'))
+async def _delete_user(db: Database, username: str, id: Optional[str]):
+    @transaction(db)
+    async def _delete(tx):
+        existing_users = await users_with_username_or_login_id(tx, username, None)
+        if len(existing_users) == 1:
+            if existing_users[0]['state'] != 'deleted':
+                where_conditions = ['username = %s']
+                where_args = [username]
+
+                if id is not None:
+                    where_conditions.append('id = %s')
+                    where_args.append(id)
+
+                n_rows = await tx.execute_update(
+                    f'''
+UPDATE users
+SET state = 'deleting'
+WHERE {' AND '.join(where_conditions)};
```

Comment:
you don't lock the users table, so the user could change from deleting to deleted before you run this query and then you'd change it back to deleting, which seems wrong to me. username is a unique key on the users table, so it seems to me that we should just do:
```
UPDATE users
SET state = 'deleting'
WHERE state != 'deleted' AND username = %s
```
that's guaranteed to be atomic, so no other operation can race in and change deleting to deleted and then you change deleted back to deleting.

We can use the updated row count to throw an unknown user error.